File: 0140_Word_Break_II/solution.cs
/*
139. Word Break II

Given a string s and a dictionary of words dict, add spaces in s to construct a sentence where each word 
is a valid dictionary word.

Return all such possible sentences.

For example, given
s = "catsanddog",
dict = ["cat", "cats", "and", "sand", "dog"].

A solution is ["cats and dog", "cat sand dog"].

*/
using System.Collections.Generic;


public class Solution140 {
    // DP,  instead of store bool for each index, store a list of string. see example as below.
    public IList<string> WordBreak(string s, ISet<string> wordDict) {
        var res = new List<string>();
        var solution = string.Empty;

        var dp = new List<string>[s.Length + 1];
        for (int i = 0; i < s.Length + 1; i++)
        {
            dp[i] = new List<string>();
        }

        for (int i = 0; i < s.Length; i++)
        {
            for (int j = 0; j <= i; j++)
            {
                if (j != 0 && dp[j].Count == 0) continue;

                var substr = s.Substring(j, i - j + 1);
                if (wordDict.Contains(substr))
                {
                    dp[i + 1].Add(substr);
                }
            }
        }

        Helper(dp, s.Length, res, solution);

        return res;
    }
    private void Helper(List<string>[] dp, int index, List<string> res, string solution)
    {
        if (index == 0)
        {
            solution = solution.Remove(solution.Length - 1);
            res.Add(solution);
            return;
        }

        foreach (var list in dp[index])
        {
            solution = list + " " + solution;
            this.Helper(dp, index - list.Length, res, solution);
            solution = solution.Substring(list.Length + 1);
        }
    }
    public IList<string> WordBreak_2(string s, ISet<string> wordDict) {
        return null;
    }

}

File: 0140_Word_Break_II/solution.py
'''
140. Word Break II

Given a string s and a dictionary of words dict, add spaces in s to construct a sentence where each word is a
valid dictionary word.

Return all such possible sentences.

For example, given
s = "catsanddog",
dict = ["cat", "cats", "and", "sand", "dog"].

A solution is ["cats and dog", "cat sand dog"].
'''


class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: Set[str]
        :rtype: List[str]
        """
        tokenDict = {}

        def dfs(s):
            ans = []
            if s in wordDict:
                ans.append(s)

            for x in range(len(s) - 1):
                prefix, suffix = s[:x + 1], s[x + 1:]
                if prefix not in wordDict:
                    continue

                if suffix in tokenDict:
                    rest = tokenDict[suffix]
                else:
                    rest = dfs(suffix)
                for n in rest:
                    ans.append(prefix + ' ' + n)
            tokenDict[s] = ans
            return ans

        return dfs(s)


    # http://www.cnblogs.com/zuoyuan/p/3760804.html
    def wordBreak_2(self, s, dict):
        Solution.res = []
        self.dfs(s, dict, '')
        return Solution.res

    def check(self, s, dict):
        dp = [False for i in range(len(s) + 1)]
        dp[0] = True
        for i in range(1, len(s) + 1):
            for k in range(0, i):
                if dp[k] and s[k:i] in dict:
                    dp[i] = True
        return dp[len(s)]

    def dfs(self, s, dict, stringlist):
        if self.check(s, dict):
            if len(s) == 0:
                Solution.res.append(stringlist[1:])
            for i in range(1, len(s) + 1):
                if s[:i] in dict:
                    self.dfs(s[i:], dict, stringlist + ' ' + s[:i])

input = "catsanddog"
dict = ["cat", "cats", "and", "sand", "dog"]

s = Solution()
print s.wordBreak(input, dict)
